embassy-stm32/src/pwr.rs

```diff
@@ -0,0 +1,100 @@
+//! Power control
+
+/// Configuration what peripherals should be running during standby or shutdown mode.
+/// It is determined by the disabled peripherals, if shutdown mode is used instead of standby
+/// mode if the chip supports it.
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct StandbyConfig {
+    // empty, we might want to disable RTC here, enable WKUP pins, ...
+}
+
+impl Default for StandbyConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+/// Enter standby or shutdown mode.
+///
+/// STM32 microcontrollers support the "standby" and sometimes the "shutdown" mode,
+/// where almost all peripherals and the core will be shut down.
+/// In most cases, this is the deepest sleep mode.
+/// If standby mode is entered, the program ends as the SRAM is not kept.
+/// After a wakeup the program starts from the start.
+/// It can be determined if the core started after a reset or standby mode.
+///
+/// Content of variables will be lost as the SRAM is disabled during standby.
+/// State has to be either stored inside the RTC peripheral,
+/// or on the flash.
+///
+/// The following peripherals are not shut down during standby and can wake up the core:
+///
+/// ## WKUP GPIO pins: (this is currently unimplemented)
+///
+/// STM32 microcontrollers have some dedicated WKUP pins that can optionally be used as an input
+/// to wake up during standby.
+/// If a pin is configured as wakup pin, it will be converted to an input with pull-down enabled.
+/// If the pin state switches to high, the core will wake up.
+///
+/// ## RTC event: (this is currently unimplemented)
+///
+/// The RTC can wake up the core at a specific time.
+/// If RTC is enabled during standby, more current is used.
+///
+/// ## IWDG
+///
+/// If the Independent Watchdog is enabled before standby,
+/// it will not be disabled and can still reset the core if it was not pet.
+/// This can also be used to wake up after and inaccurate amount of time has passed in standby mode,
+/// which has the advantage that it is in most cases less power hungry than the RTC.
+pub fn standby(_config: &StandbyConfig) {
```

Comment:
```suggestion
pub fn standby(_config: &StandbyConfig) -> ! {
```

denotes nicely that this doesn't return.

Reply:
Done, had to move `unreachable!` out of the critical section.